[COLLAGE] Add more customization to support more targets

krishnaraj36 opened the pull request:
1. Added custom cost module to provide a provision to incorporate custom cost estimator python function instead using default cost function.
     eg: cost_estimator = CustomCostEstimator(py_fn_estimator="tvm.relay.collage.opencl_cost_estimator")
           mod = CollagePartition(config, cost_estimator=cost_estimator)(mod)
2. Added provision to select BYOC fusion style for all compiler target.
     eg : config = { "relay.collage.byoc_fusion_style": ["compiler.NoFusion", "compiler.TVMFusion"]}
            ctxt = tvm.transform.PassContext(config=config)

srkreddy1238 commented:
@mbs-octoml can you take a look on this PR ?

srkreddy1238 commented on tests/python/contrib/test_clml/menangerie.py:
```diff
@@ -0,0 +1,4179 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
I see only resnet50, resnet50_16, mobilenet, mobilenet_16 are used for test cases.
You may remove other network definitions here.

krishnaraj36 replied:
Removed the unused code!

srkreddy1238 commented on tests/python/contrib/test_clml/test_adreno_collage_targets.py:
```diff
@@ -0,0 +1,400 @@
+
+"""Compares Collage with various other baselines."""
+
+# CAUTION: Requires some changes in python/tvm/autotvm/task/dispatcher.py
```
Is this caution relevant here ? i.e Do we need changes to AutoTVM ?

krishnaraj36 replied:
Its not required, I have removed the line.

srkreddy1238 reviewed (approved):
LGTM.

srkreddy1238 commented:
@tqchen Do we have moderator for Collage ?

srkreddy1238 commented:
@masahi any moderator for Collage ?

masahi commented:
Not sure and I'm not familiar with it either. But I can take care of this.

krishnaraj36 commented:
> @masahi : 
I have resolved all merge conflicts on this PR, can you please review and merge this PR.

krishnaraj36 reviewed (commented):
Conflicts got resolved

masahi commented:
Please remove typos and obvious grammar errors. I found many just looking over the code briefly. Please go through the change. 

Also remove `TODO(mbs)` stuff.

krishnaraj36 commented:
> Please remove typos and obvious grammar errors. I found many just looking over the code briefly. Please go through your change.
> 
> Also remove `TODO(mbs)` stuff.

@masahi 
I have fixed the most to typo error and removed TODO (mbs), Please review and let me know your feedback.
If it looks good, please merge the branch.

masahi commented on src/relay/collage/custom_cost_estimator.cc:
```diff
@@ -0,0 +1,60 @@
+
+/*!
+ * \file src/relay/collage/custom_cost_estimator.cc
+ * \brief A custom CostEstimator to support Alternative cost functions.
```
alternative

krishnaraj36 replied:
corrected typo error

masahi commented on src/relay/collage/custom_cost_estimator.h:
```diff
@@ -0,0 +1,68 @@
+
+/*!
+ * \file src/relay/collage/custom_cost_estimator.cc
+ * \brief A custom CostEstimator to support all different compiler cost.
```
to support target-specific cost functions

krishnaraj36 replied:
Corrected it

masahi commented on src/relay/collage/custom_cost_estimator.h:
```diff
@@ -0,0 +1,68 @@
+
+ protected:
+  /*!
+   * \brief Python implementation function register name to invoke.
```
Please revise this sentence, it is broken.

krishnaraj36 replied:
Corrected it.

masahi commented on src/relay/collage/custom_cost_estimator.h:
```diff
@@ -0,0 +1,68 @@
+namespace collage {
+
+/*!
+ * \brief A custom cost estimator which can determine the cost of a candidate based on
```
A cost estimator that uses a target-specific cost function

krishnaraj36 replied:
Corrected it.

masahi commented on tests/python/contrib/test_clml/menangerie.py:
```diff
@@ -0,0 +1,1527 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Use the models in python/tvm/relay/testing if possible

krishnaraj36 replied:
Yes we can use existing networks from testing instead additional network file. I have removed the file and used network from testing.

masahi commented on tests/python/contrib/test_clml/test_adreno_collage_targets.py:
```diff
@@ -0,0 +1,391 @@
+
+    logging.info(f"Invoking runner...")
+
+    os.system(f"python3 {runner_path}")
```
Why do you need this? Can't we wrap the boilarplate code in a function instead?

masahi replied:
That doesn't answer my question. Why not, instead of `runner_template` thing and invoke the string with `os.system`, wrap the same functionality in a normal function and call it however time you want.

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+  std::string config_key = "relay.collage.byoc_fusion_style";
+  std::string compiler_id = compiler;
+  Optional<Array<String>> byoc_configs = ctxt->GetConfig(config_key, Optional<Array<String>>());
+  if (!byoc_configs.defined()) {
```
Probably can remove `defined()`

krishnaraj36 replied:
Corrected it

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+  tvm::transform::PassContext ctxt = tvm::transform::PassContext::Current();
+  std::string config_key = "relay.collage.byoc_fusion_style";
+  std::string compiler_id = compiler;
```
unused

krishnaraj36 replied:
Removed unused code.

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+      } else if (fusion_name == "TVMFusion") {
+        byoc_fusion_style = kTVMFusionBYOCStyle;
+      } else if (fusion_name == "MaxDepthFusion") {
+        byoc_fusion_style = kArbitraryFusionBYOCStyle;
```
Why pick the name "MaxDepthFusion" over more obvious ones like "Arbitrary"?

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+  }
+  BYOCStyle byoc_fusion_style = kNoFusionBYOCStyle;
+  for (auto config_ : byoc_configs.value()) {
+    std::string byoc_str = static_cast<std::string>(config_);
```
static_cast unnecessary.

krishnaraj36 replied:
Yes, I have removed.

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
 }
 
+/*!
+ * \brief Returns the fusion style for /p compiler.
```
for the given compiler

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+  if (!byoc_configs.defined()) {
+    return DefaultBYOCFusionStyleForCompiler(compiler);
+  }
+  BYOCStyle byoc_fusion_style = kNoFusionBYOCStyle;
```
I think it is better to use the output of `DefaultBYOCFusionStyleForCompiler(compiler)` here.

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+    std::string byoc_str = static_cast<std::string>(config_);
+    std::string byoc_compiler = byoc_str.substr(0, byoc_str.find(".", 0));
+    if (byoc_compiler == compiler) {
+      std::string fusion_name = byoc_str.substr(byoc_str.find(".", 0) + 1, byoc_str.size());
```
Two `substr` look cryptic without an example of an input string. I suggest making a function to do the split and call it like

auto [compiler_name, fusion_style] = ...

BTW, is there really a need to prefix the fusion style with the compiler name?

krishnaraj36 replied:
Its good idea, I have defined generic string split function in utils.cc file, we can call direct for this purpose.

masahi commented on tests/python/contrib/test_clml/test_adreno_collage_targets.py:
```diff
@@ -0,0 +1,391 @@
+        config = {
+            "relay.collage.tvm_max_depth": TVM_MAX_DEPTH,
+            "relay.collage.byoc_max_depth": BYOC_MAX_DEPTH,
+            "relay.collage.byoc_fusion_style": ["clml.NoFusion"],
```
Why can't you just pass "NoFusion" here? (without clml prefix)

krishnaraj36 replied:
Since collage can support multiple compiler target, to define each byoc compiler target fusion style, we have defined such that
e.g., In case of NVidia target -  
byoc_fusion_style = [cutlass.NoFusion, cudnn.NoFusion, tensorrt.TVMFusion]

So, we have uniformly defined to support all such combinations.

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+  tvm::transform::PassContext ctxt = tvm::transform::PassContext::Current();
+  std::string config_key = "relay.collage.byoc_fusion_style";
+  std::string compiler_id = compiler;
+  Optional<Array<String>> byoc_configs = ctxt->GetConfig(config_key, Optional<Array<String>>());
```
Why is this an array? Isn't a single fusion style enough?

krishnaraj36 replied:
Byoc fusion style config can have multiple compiler fusion styles.
e.g. In case of NVidia target -
byoc_fusion_style = [cutlass.NoFusion, cudnn.NoFusion, tensorrt.TVMFusion]

krishnaraj36 commented:
@masahi 
Thank you for feedback.
I have added modification to code as per your suggestion and write out clarification for you few quarries.
Please review again and let me know you feedback.

masahi commented on src/relay/collage/gather_partition_specs.cc:
```diff
@@ -103,6 +101,38 @@ BYOCStyle BYOCFusionStyleForCompiler(const String& compiler) {
+    LOG(INFO) <<"byoc cfg: "<<byoc_cfg[0]<<" : "<<byoc_cfg[1];
+    if (byoc_cfg[0] == compiler) {
+      if (byoc_cfg[1] == "NoFusion") {
+        LOG(INFO)<<"NoFusion detected";
```
Remove `LOG(INFO)`, three in this file

krishnaraj36 replied:
Removed prints

masahi commented on src/relay/collage/utils.cc:
```diff
@@ -134,6 +134,19 @@ bool MustBeLowered(const Expr& expr) {
   return false;
 }
 
+std::vector<std::string> SplitString(std::string stmt, const char* del) {
```
Just define this in `gather_partition_specs.cc` where it's used, since it is a one-off function

krishnaraj36 replied:
I have defined generic split string functions, which can be useful for any kind of splitting of string statement in future. So I defined as utility function.

masahi commented on tests/python/contrib/test_clml/test_adreno_collage_targets.py:
```diff
@@ -0,0 +1,401 @@
+
+# Code to run a model. The actual call to 'run' is appended at compile time.
+# We invoke the model as a sub-process so that we can wrap profiling tools around it.
+runner_template = f"""
```
Just make it a function instead of a template.

krishnaraj36 replied:
I had followed same test script used for Nvidia test case.(tests/python/relay/collage/demo_collage_partitioner.py)

Since here we are not doing any profiling for adreno target, so we can have single function instead runner template.

I have added the changes for that.

krishnaraj36 commented:
@masahi 
Thanks for feedback, I have added the modification to the code as per your suggestion.
Please review and let me know your feedback.

krishnaraj36 reviewed (commented):
I have added the modification to the code as per your suggestion.
Please review and let me know your feedback.
